usr/lib/common/mech_openssl.c

```diff
@@ -4898,6 +4898,325 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
     return rv;
 }
 
+CK_RV calc_rsa_crt_from_me(CK_ATTRIBUTE *modulus, CK_ATTRIBUTE *pub_exp,
+                           CK_ATTRIBUTE *priv_exp, CK_ATTRIBUTE **prime1,
+                           CK_ATTRIBUTE **prime2, CK_ATTRIBUTE **exponent1,
+                           CK_ATTRIBUTE **exponent2, CK_ATTRIBUTE **coef)
+{
+    BN_CTX *bn_ctx;
+    BIGNUM *n, *e, *d, *k, *r, *t, *two, *g, *y, *n_minus_1, *j, *x;
+    BIGNUM *p, *q, *dp, *dq, *invq;
+    int i, prime_len;
+    CK_BYTE *buff = NULL;
+    CK_RV rc;
+
+    bn_ctx = BN_CTX_secure_new();
+    if (bn_ctx == NULL) {
+        TRACE_ERROR("BN_CTX_secure_new failed\n");
+        return CKR_FUNCTION_FAILED;
+    }
+
+    /* Get modulus as BIGNUM */
+    n = BN_CTX_get(bn_ctx);
+    if (n == NULL ||
+        BN_bin2bn(modulus->pValue, modulus->ulValueLen, n) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for modulus\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Get public exponent as BIGNUM */
+    e = BN_CTX_get(bn_ctx);
+    if (e == NULL ||
+        BN_bin2bn(pub_exp->pValue, pub_exp->ulValueLen, e) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for public exponent\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Get private exponent as BIGNUM */
+    d = BN_CTX_get(bn_ctx);
+    if (d == NULL ||
+        BN_bin2bn(priv_exp->pValue, priv_exp->ulValueLen, d) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for private exponent\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    n_minus_1 = BN_CTX_get(bn_ctx);
+    two = BN_CTX_get(bn_ctx);
+    k = BN_CTX_get(bn_ctx);
+    r = BN_CTX_get(bn_ctx);
+    t = BN_CTX_get(bn_ctx);
+    g = BN_CTX_get(bn_ctx);
+    y = BN_CTX_get(bn_ctx);
+    j = BN_CTX_get(bn_ctx);
+    x = BN_CTX_get(bn_ctx);
+    p = BN_CTX_get(bn_ctx);
+    q = BN_CTX_get(bn_ctx);
+    dp = BN_CTX_get(bn_ctx);
+    dq = BN_CTX_get(bn_ctx);
+    invq = BN_CTX_get(bn_ctx);
+    if (n_minus_1 == NULL || two == NULL || k == NULL || r == NULL ||
+        t == NULL || g == NULL || y == NULL || j == NULL || x == NULL ||
+        p == NULL || q == NULL || dp == NULL || dq == NULL || invq == NULL) {
+        TRACE_ERROR("BN_CTX_get failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    if (BN_set_word(two, 2) != 1 ||
+        BN_sub(n_minus_1, n, BN_value_one()) != 1) {
+        TRACE_ERROR("BN_set_word/BN_sub failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /*
+     * Prime-Factor Recovery from n, e and d described in NIST Special
+     * Publication 800-56B R1 Recommendation for Pair-Wise Key Establishment
```

Comment:
Looks like Revision 1 is withdrawn, see https://csrc.nist.gov/pubs/sp/800/56/b/r1/final
Maybe we should reference Rev2. The algorithm is the same.

Reply:
Fixed (to R2, also in the commit description)